tools/emscripten_wasmfs_fuzzer/workload.h

```diff
@@ -0,0 +1,166 @@
+  // This could be a list of files to rename or random file content to write.
+  virtual void generate() = 0;
+
+  // This function defines the actual file system property that should be
+  // validated. Any file system syscalls should be called by this function.
+  virtual void execute() = 0;
+};
+
+// This workload class attempts to fuzz for tearing in reads and writes.
+// This should validate that writes are atomic (i.e. that a read interleaved
+// between two reads should not read a portion of the first read and a portion
+// of the second read). Writer threads should write a uniform string of the same
+// character. Reader threads should validate that file content is not intermixed
+// (it should check that all characters in the file content are the same).
+class ReadWrite : public Workload {
+  int fd;
+  std::vector<std::string> work;
+  std::unordered_map<std::thread::id, std::string> threads;
+  std::atomic<bool> go{false};
+
+  bool isSame(std::vector<char>& target) {
+    std::set<char> seen;
+
+    for (auto it = target.begin(); it != target.end() - 1; it++) {
+      seen.insert(*it);
+      if (seen.size() > 1) {
+#ifdef FUZZER_DEBUG
+        for (int i = 0; i < target.size(); i++) {
+          if (target[i] == 0) {
+            printf("\\0");
+          } else {
+            printf("%c", target[i]);
+          }
+        }
+#endif
+        return false;
+      }
+    }
+    return true;
+  }
+
+  void writer() {
+    while (!go) {
+    }
+
+    for (int i = 0; i < rand.upTo(10); i++) {
+#ifdef FUZZER_DEBUG
+      std::cout << std::this_thread::get_id() << " is running.\n";
+#endif
+      auto current = rand.upTo(work.size());
+
+      int bytesWritten =
+        pwrite(fd, work[current].c_str(), work[current].size(), 0);
+    }
+  }
+
+  void reader() {
+    while (!go) {
+    }
+#ifdef FUZZER_DEBUG
+    std::cout << std::this_thread::get_id() << " is running.\n";
+#endif
+
+    int size = work.begin()->size();
+    // Add room for null terminator.
+    std::vector<char> buf(size + 1, 0);
+
+    for (int i = 0; i < rand.upTo(10); i++) {
+      std::fill(buf.begin(), buf.end(), 0);
+
+      int bytesRead = pread(fd, buf.data(), size, 0);
+#ifdef FUZZER_DEBUG
+      printf("Read the following data: %s\n", buf.data());
+#endif
+      assert(bytesRead == work.begin()->size());
+
+      assert(isSame(buf));
+    }
+  }
+
+public:
+  ReadWrite(Random& rand) : Workload(rand) {}
+
+  void generate() override {
+    // TODO: parameterize the number of input strings.
+    // This function should produce a string that consists of 1 character type.
+    auto size = rand.upTo(10) + 1;
+    work.resize(size);
+
+    auto length = rand.getNonZero();
+
+    for (int i = 0; i < work.size(); i++) {
+      work[i] = rand.getSingleSymbolString(length);
+    }
+  }
+
+  void execute() override {
+    // Generate a test file to read and write to.
+    fd = open("test", O_CREAT | O_RDWR, 0777);
+
```

Comment:
It would be good to delete this file at the end of the workload. That way state from one workload doesn't unintentionally leak into later workloads.